components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/token/AbstractJWTGenerator.java

```diff
@@ -0,0 +1,552 @@
+            }
+        }
+
+        signatureAlgorithm = ServiceReferenceHolder.getInstance().getAPIManagerConfigurationService().
+                getAPIManagerConfiguration().getFirstProperty(AppMConstants.SIGNATURE_ALGORITHM);
+        if (signatureAlgorithm == null || !(signatureAlgorithm.equals(NONE) || signatureAlgorithm.equals(
+                SHA256_WITH_RSA))) {
+            signatureAlgorithm = SHA256_WITH_RSA;
+        }
+
+        addClaimsSelectively = Boolean.parseBoolean(ServiceReferenceHolder.getInstance().
+                getAPIManagerConfigurationService().getAPIManagerConfiguration().
+                getFirstProperty(AppMConstants.API_CONSUMER_AUTHENTICATION_ADD_CLAIMS_SELECTIVELY));
+    }
+
+    /**
+     * Method that generates the JWT token from SAML2 response
+     *
+     * @param saml2Assertions
+     * @param webAppInfoDTO
+     * @param messageContext
+     * @return jwt token
+     * @throws org.wso2.carbon.appmgt.api.AppManagementException
+     */
+    public String generateToken(Map<String, Object> saml2Assertions, WebAppInfoDTO webAppInfoDTO,
+                                MessageContext messageContext) throws AppManagementException {
+
+        String endUserName = (String) saml2Assertions.get(SAMLConstants.SAML2_ASSERTION_SUBJECT);
+        String jwtHeader = buildHeader(endUserName);
+
+        String jwtBody = buildBody(saml2Assertions);
+
+        String base64EncodedHeader = Base64Utils.encode(jwtHeader.getBytes());
+        String base64EncodedBody = Base64Utils.encode(jwtBody.getBytes());
+
+        if (signatureAlgorithm.equals(SHA256_WITH_RSA)) {
+            String assertion = base64EncodedHeader + "." + base64EncodedBody;
+            /* Get the assertion signed */
+            byte[] signedAssertion = signJWT(assertion, endUserName);
+
+            if (log.isDebugEnabled()) {
+                log.debug("signed assertion value : " + new String(signedAssertion));
+            }
+            String base64EncodedAssertion = Base64Utils.encode(signedAssertion);
+            return base64EncodedHeader + "." + base64EncodedBody + "." + base64EncodedAssertion;
+        } else {
+            return base64EncodedHeader + "." + base64EncodedBody + ".";
+        }
+    }
+
+    public String buildHeader(String endUserName) throws AppManagementException {
+        String jwtHeader = null;
+        if (NONE.equals(signatureAlgorithm)) {
+            StringBuilder jwtHeaderBuilder = new StringBuilder();
+            jwtHeaderBuilder.append("{\"typ\":\"JWT\",");
+            jwtHeaderBuilder.append("\"alg\":\"");
+            jwtHeaderBuilder.append(JWTSignatureAlgorithm.NONE.getJwsCompliantCode());
+            jwtHeaderBuilder.append("\"");
+            jwtHeaderBuilder.append("}");
+            jwtHeader = jwtHeaderBuilder.toString();
+        } else if (SHA256_WITH_RSA.equals(signatureAlgorithm)) {
+            jwtHeader = addCertToHeader(endUserName);
```

Comment:
Need to refactor this method.

The initial header should be built first and then the certificate thumbprint should be added if their is an algorithm.